devito/symbolics/manipulation.py

```diff
@@ -319,12 +319,82 @@ def reuse_if_untouched(expr, args, evaluate=False):
         return expr.func(*args, evaluate=evaluate)
 
 
-def evalmin(a, b):
+def evalrel(func=min, input=None, assumptions=[]):
     """
-    Simplify min(a, b) if possible.
+    Simplify evalmin if possible and return nested MIN/MAX expression. This function
+    takes advantage of python's built-in `min`, Sympy's `Min` and our in-house `MIN`
+    to reduce multiple candidates in relation's when possible.
+
+    Parameters
+    ----------
+    func : min or max, the operation to follow
+        Defaults to min
+    input : a list of the candidate symbols to be simplified
+        The candidates to be evaluated. Defaults to None.
+    assumptions : A list of assumptions formed as relationals between candidates,
+        assumed to be True. Defaults to None.
+
+    Examples
+    --------
+    Assuming no values are known for `a`, `b`, `c`, `d` but we know that `d<=a` and
+    `c>=b` we can safely drop `a` and `c` from the candidate list.
+
+    >>> from devito import Symbol
+    >>> a = Symbol('a')
+    >>> b = Symbol('b')
+    >>> c = Symbol('c')
+    >>> d = Symbol('d')
+    >>> evalrel(max, [a, b, c, d], [[Le(d, a), Ge(c, b)]])
+    MAX(a, c)
     """
+    sfunc = (Min if func is min else Max)  # Choose sympy's Min/Max
+    rfunc = (rmin if func is min else rmax)  # Choose built-in MIN/MAX
+
+    mapper = {}
+
+    processed = []
+
+    # Form relationals so that RHS has more arguments than LHS:
+    # i.e. (a + d >= b) ---> (b <= a + d)
+    assumptions = [asm if len(asm.rhs.args) > len(asm.lhs.args) else
+                   asm.reversed for asm in assumptions]
+
+    for asm in assumptions:
+        if (asm.__class__ in (Ge, Gt)) and asm.rhs.is_Add and asm.lhs.is_positive:
+            # If `c >= a + b, {a, b, c} >= 0` then add 'c>=a, c>=a'
+            if all(i.is_positive for i in asm.rhs.args):
+                processed.extend(Ge(asm.lhs, i) for i in asm.rhs.args)
+            # If `c >= a + b, a>= 0, b<=0` then add 'c>=b, c<=a'
+            elif len(asm.rhs.args) == 2:
```

Comment:
We guarantied there is no corner case not in the `if` or `elif`? I would add a final `else` for safety.

Reply:
Of course there are. Added. Again, this is elementary support...probably gonna be enhanced along the way.